### PR TITLE
🚧 Add Multi Platform support for Targets 

### PR DESCRIPTION
**Pull request by ollieatkinson**
Resolves https://github.com/tuist/tuist/issues/397

### Short description 📝

Following the blog post from Dave De Long (https://davedelong.com/blog/2018/11/15/building-a-crossplatform-framework/) we are adding support into the manifest for defining a target which supports multiple platforms.

```swift
import ProjectDescription

let project = Project(name: "Framework2",
                      targets: [
                          Target(name: "Framework2",
                                 platform: [ .iOS, .macOS ],
                                 product: .framework,
                                 bundleId: "io.tuist.Framework2",
                                 infoPlist: "Config/Framework2-Info.plist",
                                 sources: "Sources/**",
                                 headers: Headers(public: "Sources/Public/**", 
                                                  private: "Sources/Private/**", 
                                                  project: "Sources/Project/**"),
                                 dependencies: []),

                          Target(name: "Framework2Tests",
                                 platform: [ .iOS, .macOS ],
                                 product: .unitTests,
                                 bundleId: "io.tuist.Framework2Tests",
                                 infoPlist: "Config/Framework2Tests-Info.plist",
                                 sources: "Tests/**",
                                 dependencies: [
                                     .target(name: "Framework2"),
                          ]),
])
```

### Solution 📦

- Use an OptionSet in the manifest to allow _one_ or _many_ platforms to be specified, this also means it's impossible to define duplicates.
- When generating the project settings if multiple platforms are specified ensure build settings include the `[sdk=_*]` syntax. e.g.

```
        "SWIFT_ACTIVE_COMPILATION_CONDITIONS[sdk=macosx*]": "DEBUG",
        "CODE_SIGN_IDENTITY[sdk=iphoneos*]": "iPhone Developer",
        "SDKROOT[sdk=macosx*]": "macosx",
        "COMBINE_HIDPI_IMAGES[sdk=macosx*]": "YES",
```

- Ensure the linter now checks that _any_ of the supported platforms can resolve the dependency graph.

### Implementation 👩‍💻👨‍💻

- [x] Add manifest API for platform
- [x] Change build settings per-platform
- [x] Linter for multi platform targets 
- [ ] Update documentation
- [ ] Update change log

### Test Plan 🛠:

- Generate `fixtures/ios_app_with_frameworks` via `tuist generate`
- Run the following:
```
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework1 -sdk macosx
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework2 -sdk macosx
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework1 -sdk iphoneos
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework2 -sdk iphoneos
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework1 -sdk iphonesimulator
xcrun xcodebuild -workspace Workspace.xcworkspace -scheme Framework2 -sdk iphonesimulator
```
- Ensure _all_ builds succeed
_optional_:
- Open the project in Xcode 
- Play with it

### CODE REVIEW

File: Sources/TuistGenerator/Linter/GraphLinter.swift [hunk 1]
```diff
@@ -105,20 +105,34 @@ class GraphLinter: GraphLinting {
...
-                                        product: from.target.product)
-        let toTarget = LintableTarget(platform: to.target.platform,
-                                      product: to.target.product)
+        var validLinks: [Bool] = []
 
-        if !GraphLinter.validLinks.keys.contains(fromTarget) {
-            let reason = "Target \(from.target.name) has a platform '\(from.target.platform)' and product '\(from.target.product)' invalid or not supported yet."
-            let issue = LintingIssue(reason: reason, severity: .error)
-            issues.append(issue)
+        for fromPlatform in from.target.platform {
```
[the remark this answers is not in the document]
Comment (kwridan) [reply Sources/TuistGenerator/Linter/GraphLinter.swift]: A few more cases to consider - transitive dependencies 

```
- App (iOS)
- Framework1 (iOS, macOS)
- Framework2 (macOS)

Dependencies:
- App -> Framework1
- Framework1 -> Framework2
```

this currently passes linting but will fail to build, worth us catching those.

File: Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift [hunk 2]
```diff
@@ -83,23 +83,38 @@ public final class DefaultSettingsProvider: DefaultSettingsProviding {
...
         let product = settingsHelper.settingsProviderProduct(target)
-        let platform = settingsHelper.settingsProviderPlatform(target)
         let variant = settingsHelper.variant(buildConfiguration)
-        let targetDefaultAll = BuildSettingsProvider.targetDefault(variant: .all,
-                                                                   platform: platform,
-                                                                   product: product,
-                                                                   swift: true)
-        let targetDefaultVariant = BuildSettingsProvider.targetDefault(variant: variant,
-                                                                       platform: platform,
+
```
[the remark this answers is not in the document]
Comment (ollieatkinson) [reply Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift]: We should probably reduce the settings and only provide them if the platforms give different values.

File: Sources/TuistGenerator/Graph/TargetNode.swift [hunk 3]
```diff
@@ -59,14 +59,16 @@ class TargetNode: GraphNode {
...
-        let dependencies: [GraphNode] = try target.dependencies.map {
-            try node(for: $0,
-                     path: path,
-                     name: name,
-                     platform: target.platform,
-                     cache: cache,
-                     circularDetector: circularDetector,
-                     modelLoader: modelLoader)
+        let dependencies: [GraphNode] = try target.dependencies.flatMap { dependency in
+            try target.platform.map { platform in
```
[the remark this answers is not in the document]
Comment (ollieatkinson) [reply Sources/TuistGenerator/Graph/TargetNode.swift]: This is a problem, but it should be able to resolve the SDKROOT from the target platform and fulfil the sdk paths.

RE the duplicate dependencies, that looks like a bug and we should only reference a single library

File: Sources/TuistGenerator/Generator/ConfigGenerator.swift [hunk 4]
```diff
@@ -164,8 +164,14 @@ final class ConfigGenerator: ConfigGenerating {
...
             settings["CODE_SIGN_ENTITLEMENTS"] = "$(SRCROOT)/\(entitlements.relative(to: sourceRootPath).pathString)"
         }
-        settings["SDKROOT"] = target.platform.xcodeSdkRoot
-        settings["SUPPORTED_PLATFORMS"] = target.platform.xcodeSupportedPlatforms
+
+        for platform in target.platform {
+            for supported in platform.xcodeSupportedPlatforms {
+                settings["SDKROOT[sdk=\(supported)*]"] = platform.xcodeSdkRoot
+            }
+            settings["SUPPORTED_PLATFORMS"] = platform.xcodeSupportedPlatforms
```
[the remark this answers is not in the document]
Comment (ollieatkinson) [reply Sources/TuistGenerator/Generator/ConfigGenerator.swift]: Odd, I had this with a join but must have accidentally reverted the change.

File: Sources/ProjectDescription/Platform.swift [hunk 5]
```diff
@@ -2,9 +2,17 @@ import Foundation
...
+    public init(rawValue: Int) {
+        self.rawValue = rawValue
+    }
+
+    public static let iOS = Platform(rawValue: 1 << 0)
+    public static let macOS = Platform(rawValue: 1 << 1)
+    public static let watchOS = Platform(rawValue: 1 << 2)
+    public static let tvOS = Platform(rawValue: 1 << 3)
+
+    public static let all: Platform = [.iOS, .macOS, .watchOS, .tvOS]
```
Comment (pepibumur) [inline Sources/ProjectDescription/Platform.swift]: 👏 good idea

File: Sources/ProjectDescription/Platform.swift [hunk 6]
```diff
@@ -2,9 +2,17 @@ import Foundation
 
 // MARK: - Platform
 
-public enum Platform: String, Codable {
-    case iOS = "ios"
-    case macOS = "macos"
-    case watchOS = "watchos"
-    case tvOS = "tvos"
+public struct Platform: OptionSet, Codable {
```
Comment (pepibumur) [inline Sources/ProjectDescription/Platform.swift]: TIL about `OptionSet`

File: Sources/TuistGenerator/Graph/TargetNode.swift (hunk 3 above)
Comment (pepibumur) [inline Sources/TuistGenerator/Graph/TargetNode.swift]: If the dependency is also a multi-platform target, I don't think we need to add an entry per platform.

File: Sources/TuistGenerator/Linter/GraphLinter.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (pepibumur) [reply Sources/TuistGenerator/Linter/GraphLinter.swift]: Agree with this one

**Comment by ollieatkinson** [comment]
Closing as stale, will reopen if I decide to continue this P/R

**Comment by ollieatkinson** [comment]
The couple of parts to point out which I didn’t get around to solving:
- platform specific build settings / configuration
- expression of platform specific build settings in the manifest
- linting, including transitive dependencies

**Comment by RomanTysiachnik** [comment]
Hey @ollieatkinson, any plans to reborn cross-platform frameworks? What was the main issue to complete that PR?

**Pull request closed without merging**